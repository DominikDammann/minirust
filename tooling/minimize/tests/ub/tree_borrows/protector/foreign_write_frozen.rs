//@ compile-flags: --minimize-tree-borrows

// Check that a foreign write of a protected mutable reference with frozen permission is UB.

fn foo(x: *mut i32, _frozen: &i32) {
    // (x, Reserved) -> (y, Reserved) -> (_frozen, P[Frozen])
    unsafe {
        *x = 42; // UB! Foreign Write to Protected Frozen.
    } 
}

fn main() {
    let x = &mut 31; // (x, Reserved)   
    let xraw = x as *mut i32; // (x, Reserved) 
    let y = unsafe { &mut *xraw }; // (x, Reserved) -> (y, Reserved)

    foo(xraw, y);
}
