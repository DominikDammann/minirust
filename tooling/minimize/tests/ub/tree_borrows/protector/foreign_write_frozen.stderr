fatal error: UB: Tree Borrows: a protected pointer with Frozen permission becomes Disabled
